endpoints/openrtb2/amp_auction.go

```diff
@@ -0,0 +1,230 @@
+					glog.Errorf("/openrtb2/amp Critical error unpacking targets: %v", err)
+					return
+				}
+				for key, value := range bidExt.Prebid.Targeting {
+					targets[key] = value
+				}
+			}
+		}
+	}
+
+	// Now JSONify the tragets for the AMP response.
+	ampResponse := AmpResponse{
+		Targeting: targets,
+	}
+
+	// Fixes #231
+	enc := json.NewEncoder(w)
+	enc.SetEscapeHTML(false)
+
+	// If an error happens when encoding the response, there isn't much we can do.
+	// If we've sent _any_ bytes, then Go would have sent the 200 status code first.
+	// That status code can't be un-sent... so the best we can do is log the error.
+	if err := enc.Encode(ampResponse); err != nil {
+		glog.Errorf("/openrtb2/amp Error encoding response: %v", err)
+	}
+}
+
+// parseRequest turns the HTTP request into an OpenRTB request.
+// If the errors list is empty, then the returned request will be valid according to the OpenRTB 2.5 spec.
+// In case of "strong recommendations" in the spec, it tends to be restrictive. If a better workaround is
+// possible, it will return errors with messages that suggest improvements.
+//
+// If the errors list has at least one element, then no guarantees are made about the returned request.
+func (deps *endpointDeps) parseAmpRequest(httpRequest *http.Request) (req *openrtb.BidRequest, errs []error) {
+	req = &openrtb.BidRequest{}
+	errs = nil
+
+	ampId := httpRequest.FormValue("tag_id")
+	if len(ampId) == 0 {
+		errs = []error{errors.New("AMP requests require an AMP config ID")}
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(50)*time.Millisecond)
+	defer cancel()
+
+	storedRequests, errs := deps.storedReqFetcher.FetchRequests(ctx, []string{ampId})
+	if len(errs) > 0 {
+		return nil, errs
+	}
+	if len(storedRequests) == 0 {
+		errs = []error{fmt.Errorf("No AMP config found for ID '%s'", ampId)}
+		return
+	}
+
+	// The fetched config becomes the entire OpenRTB request
+	requestJson := storedRequests[ampId]
+	if err := json.Unmarshal(requestJson, req); err != nil {
+		errs = []error{err}
+		return
+	}
+
+	// Two checks so users know which way the Imp check failed.
+	if len(req.Imp) == 0 {
+		errs = []error{fmt.Errorf("AMP config ID '%s' does not include an Imp object. One id required.", ampId)}
+		return
+	}
+	if len(req.Imp) > 1 {
+		errs = []error{fmt.Errorf("AMP config ID '%s' includes multiple Imp objects. We must have only one.", ampId)}
+		return
+	}
+
+	// Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
+	deps.setFieldsImplicitly(httpRequest, req)
+
+	// Need to ensure cache and targeting are turned on
+	extRequest := &openrtb_ext.ExtRequest{}
+	if req.Ext != nil && len(req.Ext) > 0 {
+		if err := json.Unmarshal(req.Ext, extRequest); err != nil {
+			errs = []error{err}
+			return
+		}
+	}
+
+	updateExt := false
+	// Ensure Targeting is on, if not set an empty (default) targeting object
```

Comment:
I don't think we should be doing this, for all the same reasons I listed here: https://github.com/prebid/prebid-server/pull/270#discussion_r162472816

It's not (always) easy to intuit what the user wants... and by trying, we might cover up real bugs. Better to just fail with the clearest and most thorough error message possible, so they know what went wrong and how to fix it.

Reply:
I thought I was failing with the most thorough error message possible. The only source of Imp data is the stored request, and no Imps mean no auction can take place.

The second test might be dropped, as it would block anyone from creating an AMP request for multiple Imps. Not sure if AMP can support multiple Imps in one request, my initial reading of the RTC spec suggests it is not possible. But someone determined enough might find a way to enable it.